django/test/client.py

```diff
@@ -494,26 +507,37 @@ def post(
             content_type,
             secure=secure,
             headers=headers,
+            query_params=query_params,
             **extra,
         )
 
-    def head(self, path, data=None, secure=False, *, headers=None, **extra):
+    def head(
+        self, path, data=None, secure=False, *, headers=None, query_params=None, **extra
+    ):
         """Construct a HEAD request."""
-        data = {} if data is None else data
+        if query_params and data:
+            raise ValueError("You cannot specify query_params and data together.")
+        query_params = query_params or data
+        query_params = {} if query_params is None else query_params
```

Comment:
Hm so we now have 2 kwargs that do the same thing just named differently… doesn't seem Pythonic 😅  (or Djangonic)

The forum discussion didn't really talk about what to do with get() & whether we have 2 or deprecate data or what.  I wonder whether it's worth asking @adamchainz & co.

Reply:
I was thinking about this as well. I'd be keen to deprecate it if only to make the code a little less messy and more consistent. There is a lot of code using this, but only in tests and I feel like breaking people's tests is tolerable relative to breaking their sites.

On the other hand, the current API has been there a long time and it works, even if it's a little confusing.

Reply:
I agree that we cannot have two parameters for the same. Personally, I wouldn't add `query_params` to `get()` and `head()`. I think we need to reach a wider audience and ask on the forum.

Reply:
I added a post with my rationale: https://forum.djangoproject.com/t/handling-test-client-data-query-params/25409

Reply:
Do they do the same thing? `query_params` explicitly sets the query string. `data` sets form data on the request in accordance with the HTTP rules for the method in use. (It's an implementation detail that that happens to be on the query string.)

I may want to use the `query_params` **or**  send form `data`, depending on my test. 

I just can't use **both** at the same time. 

A form like this...

```
<form method="get" action="/test-view/?a=1&b=2">
    <input type="hidden" name="c" value="3">
    <input type="hidden" name="d" value="4">
    <input type="submit" value="Submit">
</form>
```

… submits to `/test-view/?c=3&d=4`, so `get()` and `head()` should set `data` in place of `query_params` if both are provided, and raise a warning — or error if you prefer — that `query_params` cannot be used in this case. 

tl;dr: I think having both is fine. It keeps the API consistent. It's **using** both **at the same time** that we should guard against

I'd be :-1: to removing the `data` argument, as a whole ecosystem's worth of tests would need updating, and it's not worth that. (`django-upgrade` is available in theory but people have to learn about it and then use it, which is not a given...)

Reply:
Yeah but following that logic `query_params` would never get sent 🤔 because:

```
<form method="get" action="/test-view/?a=1&b=2">
    <input type="submit" value="Submit">
</form>
```

submits to `/test-view/?`

ie query params are always ignored.

which is even better argument for just having `data` 😬

Reply:
If you set `data` that's a _submitting a form_ type request, so **Don't use `query_params` here** is (would be) great!

But there's **loads** of times when I'm making `GET` requests that aren't form submissions and `query_params` is the natural argument in that case. 

(Or so it seems to me 😜)